Fix unwanted additional spaces added around pasted text on Windows

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

When pasting text internally within the editor on Windows additional spaces are added to either side of the pasted text.

* Copied text (i.e. what the user _intended_ to copy) `two`.
* Resulting text when pasted ` two `.

This appears to be due to [windows providing additional unwanted data when retrieving `text/html` from the clipboard event](https://discuss.prosemirror.net/t/space-added-on-paste/1274/7) data. The actual content copied is:

```
<html>
 <body>
   <!--StartFragment-->
   Copied text here
   <!--EndFragment-->
 </body>
</html>
```

When this passes through the editor's paste mechanics and the format API we end up with spaces around the text.

To fix this, this PR attempts to use regex to remove:

1. The two "Fragment" comments.
2. Whitespace either side of the comments.
It does not remove  whitespace inside the "Fragment" comments because the user may have intentionally copied some spaces which we would want to preserve.

Fixes https://github.com/WordPress/gutenberg/issues/33283

## How has this been tested?

* Use Windows 10 and latest MS Edge browser.
* Create a new Post.
* Write some text such as `Hello I am some text. Let's copy me`.
* Using your mouse highlight a word in that text being sure _not_ to copy any whitespace.
* Paste the copied text into the middle of another word.
* See that there are no spaces added either side of the text your pasted.
* Repeat but this time copy some text that does include some whitespace either side. We need to check that in this scenario the whitespace _is_ preserved.

Also test on Mac to ensure it doesn't regress this implementation.

## Screenshots <!-- if applicable -->

### Windows

#### Before

https://user-images.githubusercontent.com/444434/126521867-fdb613d2-6895-4490-bedd-7476fe09c83b.mov

#### After

https://user-images.githubusercontent.com/444434/126629281-2e5f0e68-7b59-4e95-b26e-a5f53ff0a632.mp4

### Mac OS

The below shows only the _after_ fix as the bug did not manifest itself on Mac even before this fix.

https://user-images.githubusercontent.com/444434/126629800-60050353-a75f-40e4-803a-57158cdacab6.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Remove the whitespace before the fragments that are provided by windows

See https://discuss.prosemirror.net/t/space-added-on-paste/1274/7

getdave committed:
Extract to util

getdave commented:
Ok @ellatrix it looks like t[his solves the reported Issue](https://github.com/WordPress/gutenberg/issues/33283#issuecomment-885492718). If you have time to review my implementation that would be great.

draganescu reviewed (commented):
Given that this is based on a documented OS specific issue that other editos encountered I say it's ✅ But we have a [paste-handler](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/raw-handling/paste-handler.js) that could handle this better than directly in Rich Text?

getdave commented:
> ...we have a [paste-handler](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/raw-handling/paste-handler.js) that could handle this better than directly in Rich Text?

🤔 Yeh actually it might be better moved to there. Let me see...

getdave commented:
@draganescu Hold up...we are already parsing out those markers in the paste handler 

https://github.com/WordPress/gutenberg/blob/055af1f61841ea8844cd0073ea6cb6a1d99ec208/packages/blocks/src/api/raw-handling/paste-handler.js#L95-L103

Why isn't it working then? 🤔

getdave commented:
Ok it looks like we go down the `isInternal` code path when pasting within the editor. That doesn't invoke the paste handler which is why we need the additional fix in this PR.

https://github.com/WordPress/gutenberg/blob/339a7d783ad53dc9d437b3b5764714843caea2b8/packages/block-editor/src/components/rich-text/use-paste-handler.js#L109-L120

@ellatrix I'd like to understand whether you agree that it's a good idea to strip the Windows specific fragment markers in all cases as I've done in this PR. If we don't do it here then it's never handled because whilst [the `pasteHandler` does deal with the fragments](https://github.com/WordPress/gutenberg/blob/055af1f61841ea8844cd0073ea6cb6a1d99ec208/packages/blocks/src/api/raw-handling/paste-handler.js#L95-L103) in this "internal" paste scenario it is never invoked. As a result Windows exhibits this bug.

getdave committed:
Avoid creating public APIs for util and improve function naming for clarity

As we're not sure about this fix let's not create a public API via utils. Instead revert to inlining.

Also renaming the function to make it's purpose clearer. It's not a generic normalizing function so let's make it obvious this is Windows specific.

draganescu reviewed (approved):
I think this is a good fix and despite my doubts about the place where the fix is applied we can always refactor this later.

kevin940726 reviewed (commented):
I tested it in Edge. Sometimes when you copy more than one words (with multiple spaces in between) and paste it somewhere, the copied texts are messed up, in a unpredictable way. Are you able to reproduce it?

getdave commented:
> I tested it in Edge. Sometimes when you copy more than one words (with multiple spaces in between) and paste it somewhere, the copied texts are messed up, in a unpredictable way. Are you able to reproduce it?

No I'm not. 

https://user-images.githubusercontent.com/444434/128869565-25917d87-428c-4be2-935b-3f15fa4f42ab.mov

Are you able to provide exact steps to reproduce?

getdave commented:
@kevin940726 Let me know if can no longer replicate this. If you can't then I think we should look to merge the fix as it's been around for a while now and folk are waiting on the PR to land.

kevin940726 commented on packages/block-editor/src/components/rich-text/use-paste-handler.js:
```diff
@@ -222,3 +225,17 @@ export function usePasteHandler( props ) {
+ * @return {string} the normalized html
+ */
+function removeWindowsFragments( html ) {
+	const startReg = new RegExp( '.*<!--StartFragment-->', 's' );
```
Nitpick: I think we prefer to use regex literal when creating regex?

```suggestion
	const startReg = /.*<!--StartFragment-->/s;
```

getdave replied:
Fixed

kevin940726 reviewed (approved):
I can't reproduce the bug anymore, let's ship it 👍

getdave committed:
Make both regexes literal

ellatrix commented:
Could you please add an integration test?

kevin940726 commented:
I believe the bug is only reproducible on Windows though, so not sure if we have the established testing environment yet.

ellatrix commented:
@kevin940726 A lot of the integration tests we have are for Windows. We just need a copy of the HTML that was inserted. See https://github.com/WordPress/gutenberg/tree/trunk/test/integration/fixtures/documents.

Please add an integration test because issues like this are easy to regress with refactors.

ellatrix commented:
See also https://github.com/WordPress/gutenberg/blob/trunk/test/integration/blocks-raw-handling.test.js

kevin940726 commented:
@ellatrix Cool! Didn't know that! So it's more like a unit/integration test. For some reasons I was thinking e2e tests 😅. @getdave Care to create a follow-up PR? Or maybe someone else can do that.

getdave commented:
Greetings 👋 

Yes good point. I was thinking along same lines as Kai with the Windows specific nature of the issue, but Ella's information puts a different spin on it.

I'll put it on my list to add a test.
